### PR TITLE
Algorithm agility and ciphersuites (Issue #95)

### PR DESCRIPTION
**Pull request by beurdouche**
DO NOT MERGE: Work in Progress

Discussion: Is this how we should define a ciphersuite -> (KEM,SYM,SIG,HASH) ?

### CODE REVIEW

File: draft-ietf-mls-protocol.md [hunk 1]
```diff
@@ -662,17 +662,113 @@ Cryptographic algorithms are indicated using the following types:
 
 ~~~~~
 enum {
+    /* Elliptic Curve Groups (ECDHE) */
```
Comment (bifurcation) [inline draft-ietf-mls-protocol.md]: If we're going to decouple like this, then we should just incorporate the TLS algorithm identifiers by reference.

File: draft-ietf-mls-protocol.md [hunk 2]
```diff
@@ -662,17 +662,113 @@ Cryptographic algorithms are indicated using the following types:
...
+  curve as defined in ANSI X9.62 {{X962}} and FIPS 186-4 {{DSS}}, and the
+  corresponding hash algorithm as defined in {{!SHS}}. The signature is
+  represented as a DER-encoded {{X690}} ECDSA-Sig-Value structure.
+
+EdDSA algorithms
+: Indicates a signature algorithm using EdDSA as defined in
+  {{RFC8032}} or its successors. Note that these correspond to the
+  "PureEdDSA" algorithms and not the "prehash" variants.
+
+RSASSA-PSS PSS algorithms
```
Comment (bifurcation) [inline draft-ietf-mls-protocol.md]: Let's not do RSA-PSS.  I don't think anyone is asking for it.

**Comment by bifurcation** [comment]
In addition to the two comments inline, note that there's an additional complication here due to `Derive-Key-Pair` -- if we do this decoupling, then we need to also have a requirement that the hash function in use produces enough entropy for the curve in use.  E.g., you wouldn't want to use Ed448 with SHA-256.

**Comment by katrielalex** [comment]
I am trying to fix up the CI and it looks like this PR is actually broken:

```
4:13:00 PM: Error: Unable to validate the XML document: draft-ietf-mls-protocol.xml
4:13:00 PM:  <string>: Line 726: IDREF attribute target references an unknown ID "SHS"
4:13:00 PM:  <string>: Line 736: IDREF attribute target references an unknown ID "SHS"
4:13:00 PM:  <string>: Line 788: IDREF attribute target references an unknown ID "SHS"
4:13:00 PM:  <string>: Line 730: IDREF attribute target references an unknown ID "RFC8032"
4:13:00 PM:  <string>: Line 694: IDREF attribute target references an unknown ID "RFC7919"
4:13:00 PM:  <string>: Line 733: IDREF attribute target references an unknown ID "RFC8017"
4:13:00 PM:  <string>: Line 724: IDREF attribute target references an unknown ID "ECDSA"
4:13:00 PM:  <string>: Line 691: IDREF attribute target references an unknown ID "DSS"
4:13:00 PM:  <string>: Line 725: IDREF attribute target references an unknown ID "DSS"
4:13:00 PM:  <string>: Line 727: IDREF attribute target references an unknown ID "X690"
4:13:00 PM:  <string>: Line 788: IDREF attribute target references an unknown ID "RFC7539"
4:13:00 PM: make: *** [draft-ietf-mls-protocol.htmltmp] Error 1
```

File: draft-ietf-mls-protocol.md (hunk 2 above)
Comment (beurdouche) [reply draft-ietf-mls-protocol.md]: I actually agree with this and will remove it. I liked the idea of aligning on TLS but randomness issues related to this are not something we want.

**Comment by beurdouche** [comment]
@bifurcation I'll try to work a bit more on this tomorrow. Anything controversial or missing ?

**Comment by beurdouche** [comment]
Specifically, the next two subsections just before "Credentials" should probably be removed or merged.

**Comment by bifurcation** [comment]
I think this one is well stale by now.  Still an issue that merits consideration, but needs a fresh PR.

**Pull request closed without merging**